src/nrniv/nrncore_write/io/nrncore_io.cpp

```diff
@@ -215,7 +215,7 @@ void write_nrnthread(const char* path, NrnThread& nt, CellGroup& cg) {
             writeint(nodeindices, n);
         }
         writedbl(data, n * sz);
-        if (nrn_is_artificial_[type]) {
+        if (data) {
             delete[] data;
```

Comment:
Here, we are saying `"delete data if a given mechanism type is an artificial cell"`. 

Given the implementation and it's usage, I don't see why this was condition was written like this. @nrnhines : anything obvious I am missing here?

@1uc : Is it this change that you said you saw unpredictable segfault? (or, could be unrelated to this exact change here?)

Reply:
I don't see how anything can know about the allocation that's being freed. This fix released a lot of memory, about 150 GB or so.

I think it could be one of the following:
* Something does get a hold of the pointer, e.g. by injecting a `writedbl_` somehow.
* We also released the stable identifiers in the "defered deletion vector". (Kinda a strong candidate.) I'm rerunning to rule this one out.
* A variant of the above, the EventQueues are freed after shrinking the pools. Hence, I free the stable identifiers before clearing the event queues. Maybe something in the event queues needs to know that something got deallocated. (And we prevent it from knowing by clearing the "defered deletion vector" too early.
* It could be an existing fault that just never segfaulted because under the precise circumstances where it happens we used to hog 50% of the memory.
* Something can tell that we're shrinking the `soa` containers. Also a change that was introduced just before adding this change to the list of modifications.

The precise neuron setup when it failed is:
https://github.com/neuronsimulator/nrn/pull/2455/commits

Reply:
I think that this fix is correct.

I believe that the relevant history is that artificial cells used to require an extra consolidation step on the NEURON side to put their data in the format that CoreNEURON expected, but now that step is done for non-artificial cells too.

It looks like `contiguous_art_data` is now unused, and could be removed.